File: Assets/src/model/enemies/Scorpion.cs
using System;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Chases you.\nAttacks and moves twice.")]
public class Scorpion : AIActor {
  public static new ActionCosts StaticActionCosts = new ActionCosts(Actor.StaticActionCosts) {
    [ActionType.MOVE] = 0.5f,
    [ActionType.ATTACK] = 0.5f,
  };

  protected override ActionCosts actionCosts => StaticActionCosts;
  public Scorpion(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 5;
  }

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        return new ChaseTargetTask(this, player);
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }

  internal override (int, int) BaseAttackDamage() => (1, 1);
}

File: web/src/model/enemies/Scorpion.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { ChaseTargetTask } from '../tasks/ChaseTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { ActionCosts } from '../BaseAction';
import { ActionType } from '../../core/types';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { entityRegistry } from '../../generator/entityRegistry';

/**
 * Fast enemy: moves and attacks at half cost (acts twice per turn).
 * Port of C# Scorpion.cs.
 */
export class Scorpion extends AIActor {
  protected get actionCosts(): ActionCosts {
    return new ActionCosts([
      [ActionType.MOVE, 0.5],
      [ActionType.ATTACK, 0.5],
      [ActionType.WAIT, 1],
      [ActionType.GENERIC, 1],
    ]);
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.hp = this._baseMaxHp = 2;
  }

  baseAttackDamage(): [number, number] {
    return [1, 1];
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        return new AttackTask(this, player);
      }
      return new ChaseTargetTask(this, player);
    }
    return new MoveRandomlyTask(this);
  }
}

entityRegistry.register('Scorpion', Scorpion);
